### PR TITLE
General: update our PHP version requirements to PHP 7.0

### LINKED ISSUE #31714: General: drop support for PHP 5
Starting with WordPress 6.3, scheduled to be released in a few weeks, the new minimum supported version of PHP will be 7.0.0.

Since we support WP -1, we should still support PHP 5.6 for a few months, but when preparing for the WordPress 6.4 release compatibility, we should be able to remove all our tests, checks, and backwards compatibility measures for PHP 5.6.

Related links:
- https://core.trac.wordpress.org/changeset/56141
- https://make.wordpress.org/core/2023/07/05/dropping-support-for-php-5/

Related Primary issue: #32865

### PR DESCRIPTION
**Pull request by jeherve**
Fixes #31714

## Proposed changes:

Now that WordPress 6.4 is available, our plugins will be able to drop support for WordPress 6.2 and require WordPress 6.3 (to happen in a follow-up PR, #34127).

WordPress 6.3 introduced a new PHP version requirement: PHP 7.0. We can now match that requirement in all our tools.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

* #32865

## Does this pull request change what data or activity we track or use?

* No

## Testing instructions:

* Check if CI is happy
* Check that the changes to the codesniffer to remove one set of rules makes sense.

### CODE REVIEW

File: docs/monorepo.md [hunk 1]
```diff
@@ -210,15 +210,15 @@ If a project contains PHP tests (typically PHPUnit), it must define `.scripts.te
...
+Tests are run with a variety of supported PHP versions from 7.0 to 8.2. If you have tests that only need to be run once, run them when `PHP_VERSION` matches that in `.github/versions.sh`.
 
 #### PHP tests for non-plugins
 
 For all project types other than WordPress plugins, the necessary version of PHPUnit and/or any other tools should be pulled in via Composer.
 
 We currently make use of the following packages in testing; it's encouraged to use these rather than introducing other tools that serve the same purpose.
 
-* [yoast/phpunit-polyfills](https://packagist.org/packages/yoast/phpunit-polyfills) supplies polyfills for compatibility with PHPUnit 5.7 to 9.0, to support PHP 5.6 to 8.0.
+* [yoast/phpunit-polyfills](https://packagist.org/packages/yoast/phpunit-polyfills) supplies polyfills for compatibility with PHPUnit 5.7 to 9.0, to support PHP 7.0 to 8.2.
```
Comment (anomiex) [inline docs/monorepo.md]:
```suggestion
* [yoast/phpunit-polyfills](https://packagist.org/packages/yoast/phpunit-polyfills) supplies polyfills for compatibility with PHPUnit 6.5 to 9.0, to support PHP 7.0 to 8.2.
```

File: projects/packages/changelogger/changelog/update-php-requirements [hunk 2]
```diff
@@ -0,0 +1,4 @@
+Significance: minor
```
Comment (anomiex) [inline projects/packages/changelogger/changelog/update-php-requirements]: Dropping support for an old version should be a major bump, as it's a BC break for anyone still using the old version.

I'm tempted to suggest leaving changelogger as-is for the moment though. Bumping this should also involve checking whether any of the other dependencies can be updated (e.g. looks like wikimedia/at-ease ^2.0 will cover PHP 7+) and removing `projects/packages/changelogger/src/php5/`.

File: projects/packages/changelogger/src/CommandLoader.php [hunk 3]
```diff
@@ -2,7 +2,7 @@
 /**
  * Compatibility stub for Symfony 6 changes.
  *
- * Symfony 6 (for PHP 8.0+) added return type hints to its interface. But we still support PHP 5.6, which doesn't recognize that syntax.
+ * Symfony 6 (for PHP 8.0+) added return type hints to its interface. But we still support PHP 7.0, which doesn't recognize that syntax.
```
Comment (anomiex) [inline projects/packages/changelogger/src/CommandLoader.php]: PHP 7.0 does recognize the type hints.

Really at this point we don't need the stub anymore, we can just copy the code from the php7 dir here. Maybe even merge in the CommandLoaderBase class to it too.

File: projects/packages/changelogger/src/php7/.phpcs.dir.phpcompatibility.xml [hunk 4]
```diff
@@ -1,4 +1,4 @@
 <?xml version="1.0"?>
 <ruleset>
-	<rule ref="Jetpack-Compat-70" />
+	<rule ref="Jetpack-Compat-71" />
```
Comment (anomiex) [inline projects/packages/changelogger/src/php7/.phpcs.dir.phpcompatibility.xml]: This is wrong. There's no need for any special config here now, and really no need for the stub that now unconditionally loads the file here rather than the PHP 5 version.

File: projects/packages/codesniffer/README.md [hunk 5]
```diff
@@ -28,7 +28,7 @@ You should then include the Jetpack rules in your `.phpcs.xml.dist`, like
 You will also likely want to set some configuration for other included rulesets:
 ```xml
 <config name="minimum_supported_wp_version" value="5.4" />
```
Comment (anomiex) [inline projects/packages/codesniffer/README.md]: Should probably bump this to at least 6.3 too.

File: projects/packages/codesniffer/README.md [hunk 6]
```diff
@@ -59,7 +59,7 @@ To assist with this case, we provide additional rulesets that disable PHPCompati
 * `<rule ref="Jetpack-Compat-81" />` will disable rules detecting issues that only apply to PHP < 8.1.
```
Comment (anomiex) [inline projects/packages/codesniffer/README.md:58]: You forgot to delete line 53 above (talking about Jetpack-Compat-70)

File: projects/packages/codesniffer/build-compat-rulesets.sh [hunk 7]
```diff
@@ -93,7 +93,7 @@ for V in "${PHP_VERSIONS[@]:1}"; do
 	cat > "Jetpack-Compat-$VV/ruleset.xml" <<EOF
 <?xml version="1.0"?>
 <ruleset name="Jetpack-Compat-$VV">
-	<description>Standard to disable PHPCompatibility rules that trigger with 5.6 but not $V.</description>
+	<description>Standard to disable PHPCompatibility rules that trigger with 7.0 but not $V.</description>
```
Comment (anomiex) [inline projects/packages/codesniffer/build-compat-rulesets.sh]:
```suggestion
	<description>Standard to disable PHPCompatibility rules that trigger with $V0 but not $V.</description>
```

File: projects/packages/codesniffer/build-compat-rulesets.sh
Comment (anomiex) [inline projects/packages/codesniffer/build-compat-rulesets.sh:1]: Good on updating the script. But you should have actually run it, it will make some changes to `projects/packages/codesniffer/Jetpack-Compat-NoWP/ruleset.xml` in addition to what you did above.

File: projects/packages/codesniffer/changelog/update-php-requirements [hunk 8]
```diff
@@ -0,0 +1,4 @@
+Significance: minor
```
Comment (anomiex) [inline projects/packages/codesniffer/changelog/update-php-requirements]: Again, major version bump.

File: projects/packages/codesniffer/changelog/update-php-requirements [hunk 9]
```diff
@@ -0,0 +1,4 @@
...
+Type: changed
+
+General: updated PHP requirement to PHP 7.0+
```
Comment (anomiex) [inline projects/packages/codesniffer/changelog/update-php-requirements]: I'd have done something more like this
```suggestion
BREAKING: Drop support for PHP 5.6. `testVersion` should be set to `7.0-`.
```

File: projects/packages/codesniffer/tests/php/tests/files/default.xml [hunk 10]
```diff
@@ -2,7 +2,7 @@
 <ruleset>
 
 	<config name="minimum_supported_wp_version" value="6.0" />
```
Comment (anomiex) [inline projects/packages/codesniffer/tests/php/tests/files/default.xml]: Again, I'd bump all these to at least 6.3 too.

File: projects/packages/my-jetpack/changelog/update-php-requirements [hunk 11]
```diff
@@ -0,0 +1,4 @@
+Significance: minor
```
Comment (anomiex) [inline projects/packages/my-jetpack/changelog/update-php-requirements]: This one could go two ways.

On the one hand we're only changing some tests, no change yet to the package itself. So in that sense it should be a patch bump, and no real need for a change entry.

On the other hand, possibly we should do a **major** bump for everything in `projects/packages/` to call out that we're no longer supporting PHP 5.6, even if there are no immediate breaking changes to any particular package.

File: projects/plugins/backup/changelog/update-php-requirements [hunk 12]
```diff
@@ -0,0 +1,4 @@
+Significance: minor
```
Comment (anomiex) [inline projects/plugins/backup/changelog/update-php-requirements]: As above, this should probably be a major bump since it's dropping support for old PHP versions. OTOH, I'm not sure just how closely we care to follow semver in this respect, I'm sure we haven't been doing major bumps every time we dropped support for old WordPress versions (even though by a strict reading we probably should have been doing it for those too).

(Yeah, this doesn't actually matter for backup, jetpack, or videopress since they use wordpress versioning rather than semver, but we may as well set the significance right for all the plugins)

File: projects/packages/codesniffer/tests/php/tests/files/i18n.php.report [hunk 13]
```diff
@@ -22,4 +22,5 @@
  64 | ERROR | [x] Mismatched text domain. Expected 'domain' but got 'wrongdomain'. (Jetpack.Functions.I18n.TextDomainMismatch)
  65 | ERROR | [x] Mismatched text domain. Expected 'domain' but got 'wrongdomain'. (Jetpack.Functions.I18n.TextDomainMismatch)
  72 | ERROR | [x] Mismatched text domain. Expected 'domain' but got 'wrongdomain'. (Jetpack.Functions.I18n.TextDomainMismatch)
- 77 | ERROR | [ ] The domain must not contain interpolated variables. Found $domain. (Jetpack.Functions.I18n.InterpolatedVariable)
+ 77 | ERROR | [ ] The domain must not contain interpolated variables. Found $domain.
+    |       |     (Jetpack.Functions.I18n.InterpolatedVariable)
```
Comment (anomiex) [inline projects/packages/codesniffer/tests/php/tests/files/i18n.php.report:26]: I had started on a PR to update codesniffer myself, but I got sidetracked by seeing this unnecessary change and figuring out the details to file https://github.com/squizlabs/PHP_CodeSniffer/issues/3924 😀

File: projects/plugins/backup/changelog/update-php-requirements (hunk 12 above)
Comment (jeherve) [reply projects/plugins/backup/changelog/update-php-requirements]: 👍 I switched to a major version bump in 6caa55c72efd1f4cd8352fb39b30055590d6efbc

File: projects/packages/my-jetpack/changelog/update-php-requirements (hunk 11 above)
Comment (jeherve) [reply projects/packages/my-jetpack/changelog/update-php-requirements]: 👍 I switched to a major version bump in 6caa55c72efd1f4cd8352fb39b30055590d6efbc

File: projects/packages/codesniffer/changelog/update-php-requirements (hunk 9 above)
Comment (jeherve) [reply projects/packages/codesniffer/changelog/update-php-requirements]: Done in 6caa55c72efd1f4cd8352fb39b30055590d6efbc

File: projects/packages/codesniffer/changelog/update-php-requirements (hunk 8 above)
Comment (jeherve) [reply projects/packages/codesniffer/changelog/update-php-requirements]: 👍 I switched to a major version bump in 6caa55c72efd1f4cd8352fb39b30055590d6efbc

File: projects/packages/changelogger/changelog/update-php-requirements (hunk 2 above)
Comment (jeherve) [reply projects/packages/changelogger/changelog/update-php-requirements]: 👍 I switched to a major version bump in 6caa55c72efd1f4cd8352fb39b30055590d6efbc

File: projects/packages/changelogger/src/CommandLoader.php (hunk 3 above)
Comment (jeherve) [reply projects/packages/changelogger/src/CommandLoader.php]: That makes sense. Is c59548eb338e3f76d0a4473d5f2158e4d62ca8a5 what you had in mind?

File: projects/packages/changelogger/src/php7/.phpcs.dir.phpcompatibility.xml (hunk 4 above)
Comment (jeherve) [reply projects/packages/changelogger/src/php7/.phpcs.dir.phpcompatibility.xml]: This should be handled in c59548eb338e3f76d0a4473d5f2158e4d62ca8a5

File: projects/packages/codesniffer/README.md (hunk 5 above)
Comment (jeherve) [reply projects/packages/codesniffer/README.md]: Done in 51746059f9565019d372d6bcfe6dce647076ea8e

File: projects/packages/codesniffer/tests/php/tests/files/default.xml (hunk 10 above)
Comment (jeherve) [reply projects/packages/codesniffer/tests/php/tests/files/default.xml]: Done in 51746059f9565019d372d6bcfe6dce647076ea8e

File: projects/packages/codesniffer/README.md (hunk 6 above)
Comment (jeherve) [reply projects/packages/codesniffer/README.md:58]: Done in 695714f019c25dc4f055bb1513c43e73866bc4ea

File: projects/plugins/beta/changelog/update-php-requirements
Comment (anomiex) [inline projects/plugins/beta/changelog/update-php-requirements:1]: No change file for "General: updated PHP requirement to PHP 7.0+" for this plugin?

File: projects/plugins/crm/changelog/update-php-requirements
Comment (anomiex) [inline projects/plugins/crm/changelog/update-php-requirements:1]: No change file for "General: updated PHP requirement to PHP 7.0+" for this plugin?

File: projects/plugins/debug-helper/changelog/update-php-requirements
Comment (anomiex) [inline projects/plugins/debug-helper/changelog/update-php-requirements:1]: No change file for "General: updated PHP requirement to PHP 7.0+" for this plugin?

File: projects/plugins/mu-wpcom-plugin/changelog/update-php-requirements
Comment (anomiex) [inline projects/plugins/mu-wpcom-plugin/changelog/update-php-requirements:1]: No change file for "General: updated PHP requirement to PHP 7.0+" for this plugin?

File: projects/packages/codesniffer/build-compat-rulesets.sh
Comment (jeherve) [reply projects/packages/codesniffer/build-compat-rulesets.sh:1]: It doesn't seem to generate or update rules when I run it right now. Did I miss something?

File: projects/plugins/jetpack/changelog/update-php-requirements [hunk 14]
```diff
@@ -0,0 +1,4 @@
+Significance: patch
```
Comment (anomiex) [inline projects/plugins/jetpack/changelog/update-php-requirements]: Missed this one. Not that it makes a difference, but just for consistency,
```suggestion
Significance: major
```

**Comment by anomiex** [comment]
This should be updated too
https://github.com/Automattic/jetpack/blob/695714f019c25dc4f055bb1513c43e73866bc4ea/projects/plugins/vaultpress/vaultpress.php#L19

File: projects/packages/changelogger/src/CommandLoader.php (hunk 3 above)
Comment (anomiex) [reply projects/packages/changelogger/src/CommandLoader.php]: Yes. Except since it's a major version bump anyway, let's just delete CommandLoaderBase.php.

File: projects/packages/codesniffer/build-compat-rulesets.sh
Comment (anomiex) [reply projects/packages/codesniffer/build-compat-rulesets.sh:1]: When I run it, it produces this diff
```diff
diff --git a/projects/packages/codesniffer/Jetpack-Compat-NoWP/ruleset.xml b/projects/packages/codesniffer/Jetpack-Compat-NoWP/ruleset.xml
index ca06e0e2cd..d7f39c7119 100644
--- a/projects/packages/codesniffer/Jetpack-Compat-NoWP/ruleset.xml
+++ b/projects/packages/codesniffer/Jetpack-Compat-NoWP/ruleset.xml
@@ -3,11 +3,8 @@
 	<description>Standard to re-enable PHPCompatibility rules disabled by PHPCompatibilityWP.</description>
 	<!-- This standard is automatically generated by build-compat-rulesets.sh -->
 
-	<rule ref="PHPCompatibility.Classes.NewClasses.errorFound" />
 	<rule ref="PHPCompatibility.Classes.NewClasses.sodiumexceptionFound" />
-	<rule ref="PHPCompatibility.Classes.NewClasses.typeerrorFound" />
 	<rule ref="PHPCompatibility.Constants.NewConstants.imagetype_webpFound" />
-	<rule ref="PHPCompatibility.Constants.NewConstants.img_webpFound" />
 	<rule ref="PHPCompatibility.Constants.NewConstants.sodium_base64_variant_originalFound" />
 	<rule ref="PHPCompatibility.Constants.NewConstants.sodium_base64_variant_original_no_paddingFound" />
 	<rule ref="PHPCompatibility.Constants.NewConstants.sodium_base64_variant_urlsafeFound" />
@@ -96,8 +93,6 @@
 	<rule ref="PHPCompatibility.FunctionUse.NewFunctions.array_key_lastFound" />
 	<rule ref="PHPCompatibility.FunctionUse.NewFunctions.is_countableFound" />
 	<rule ref="PHPCompatibility.FunctionUse.NewFunctions.is_iterableFound" />
-	<rule ref="PHPCompatibility.FunctionUse.NewFunctions.random_bytesFound" />
-	<rule ref="PHPCompatibility.FunctionUse.NewFunctions.random_intFound" />
 	<rule ref="PHPCompatibility.FunctionUse.NewFunctions.sodium_addFound" />
 	<rule ref="PHPCompatibility.FunctionUse.NewFunctions.sodium_base642binFound" />
 	<rule ref="PHPCompatibility.FunctionUse.NewFunctions.sodium_bin2base64Found" />
```

File: projects/packages/codesniffer/tests/php/tests/files/i18n-text_domain-not-set.php.ruleset.xml [hunk 15]
```diff
@@ -2,7 +2,7 @@
 <ruleset>
 
 	<config name="minimum_supported_wp_version" value="6.0" />
```
Comment (anomiex) [inline projects/packages/codesniffer/tests/php/tests/files/i18n-text_domain-not-set.php.ruleset.xml]: Missed one in 51746059f9565019d372d6bcfe6dce647076ea8e
```suggestion
	<config name="minimum_supported_wp_version" value="6.3" />
```

File: projects/packages/changelogger/src/CommandLoader.php [hunk 16]
```diff
@@ -1,18 +1,98 @@
...
- * Symfony 6 (for PHP 8.0+) added return type hints to its interface. But we still support PHP 5.6, which doesn't recognize that syntax.
- * Since specifying a return type when the interface doesn't is ok, use the version that always does that for PHP 7+ instead of figuring
- * out how to check the actual symfony version.
+ * Command loader for the changelogger tool CLI.
  *
  * @package automattic/jetpack-changelogger
  */
 
-namespace Automattic\Jetpack\Changelogger;
+namespace Automattic\Jetpack\Changelogger\php7;
```
Comment (anomiex) [inline projects/packages/changelogger/src/CommandLoader.php]:
```suggestion
namespace Automattic\Jetpack\Changelogger;
```

File: projects/plugins/mu-wpcom-plugin/changelog/update-php-requirements
Comment (jeherve) [reply projects/plugins/mu-wpcom-plugin/changelog/update-php-requirements:1]: Done in 3c6a8241b71f5e4f994fb1e6a6cb3e95700c2784

File: projects/plugins/debug-helper/changelog/update-php-requirements
Comment (jeherve) [reply projects/plugins/debug-helper/changelog/update-php-requirements:1]: Done in 3c6a8241b71f5e4f994fb1e6a6cb3e95700c2784

File: projects/plugins/crm/changelog/update-php-requirements
Comment (jeherve) [reply projects/plugins/crm/changelog/update-php-requirements:1]: Done in 3c6a8241b71f5e4f994fb1e6a6cb3e95700c2784

File: projects/plugins/beta/changelog/update-php-requirements
Comment (jeherve) [reply projects/plugins/beta/changelog/update-php-requirements:1]: Done in 3c6a8241b71f5e4f994fb1e6a6cb3e95700c2784

File: projects/packages/codesniffer/build-compat-rulesets.sh
Comment (jeherve) [reply projects/packages/codesniffer/build-compat-rulesets.sh:1]: Done in e04ee2dcbf61dd2383f34451cf2f94906e6009f6

File: projects/packages/changelogger/src/CommandLoader.php (hunk 3 above)
Comment (jeherve) [reply projects/packages/changelogger/src/CommandLoader.php]: Done in dfd024f1218c576e931fe46324c7e3e034123132

**Comment by jeherve** [comment]
> This should be updated too
> 
> https://github.com/Automattic/jetpack/blob/695714f019c25dc4f055bb1513c43e73866bc4ea/projects/plugins/vaultpress/vaultpress.php#L19

I'd rather not do that here, to be honest. The VaultPress plugin still requires WordPress 5.2:
https://github.com/Automattic/jetpack/blob/13612b1618f4620854d75b7c642df8947b696817/projects/plugins/vaultpress/readme.txt#L4

I'm thinking we should update that before we up the PHP requirement.

I do not recall why we haven't updated the WordPress requirement in the past, but I assume there is a reason. We last updated it before we brought it to the monorepo, in https://github.com/Automattic/vaultpress/pull/101/

I couldn't find anything on p2 either. Maybe worth doing in a follow-up PR?

**Comment by anomiex** [comment]
These should match. If you'd rather leave them both at 5.6 for now, I could go with that.
https://github.com/Automattic/jetpack/blob/5519f4c49591715b27c0a9b9be24b435e7104a2e/projects/plugins/vaultpress/readme.txt#L7 https://github.com/Automattic/jetpack/blob/5519f4c49591715b27c0a9b9be24b435e7104a2e/projects/plugins/vaultpress/vaultpress.php#L19

**Comment by jeherve** [comment]
> These should match. If you'd rather leave them both at 5.6 for now, I could go with that.
> 
> https://github.com/Automattic/jetpack/blob/5519f4c49591715b27c0a9b9be24b435e7104a2e/projects/plugins/vaultpress/readme.txt#L7
> 
> https://github.com/Automattic/jetpack/blob/5519f4c49591715b27c0a9b9be24b435e7104a2e/projects/plugins/vaultpress/vaultpress.php#L19

Ah yeah, I didn't mean to replace that one, thank you! 82cbf5098d87f179e7dc6f37d49e4f43ae4abb50

File: projects/plugins/vaultpress/changelog/update-php-requirements
Comment (anomiex) [inline projects/plugins/vaultpress/changelog/update-php-requirements:1]: This change is no longer happening, delete this file. Then the plugin's version will need to be un-bumped too.

**Review by anomiex — APPROVED** [review]
Let's do it!
